File: experiments/020-PythonVsJava/cells/python-pytest_1-Empty_free/2022-1c-recuperatorio-parcial1/20260921-213435-24827/analysis/acceptance-project/src/customerimporter/identification.py
from abc import abstractmethod

from customerimporter.typed_value import TypedValue


class Identification(TypedValue):

    # instance creation

    @classmethod
    def for_(cls, an_identification_type, an_identification_number):
        return cls._type_representing(an_identification_type)(an_identification_number)

    @classmethod
    def can_represent(cls, an_identification_type):
        return cls.type_code() == an_identification_type

    @classmethod
    @abstractmethod
    def type_code(cls):
        pass

    # error messages

    @classmethod
    def invalid_type_error_description(cls):
        return "Invalid identification type"

    # identification

    def is_for(self, an_identification_type, an_identification_number):
        return self.type_code() == an_identification_type and self.number() == an_identification_number

    def number(self):
        return self._value

    # cuit

    def cuit_number_if_none(self, a_none_closure):
        return a_none_closure()

    def has_cuit_as_identification(self):
        return False

    # dni

    def dni_number_if_none(self, a_none_block):
        return a_none_block()

    def has_dni_as_identification(self):
        return False

File: experiments/020-PythonVsJava/cells/java-gradle_1-Empty_free/2022-1c-recuperatorio-parcial1/20260921-204311-76559/workdir/src/main/java/customerimporter/Identification.java
package customerimporter;

import java.util.Map;
import java.util.Optional;
import java.util.function.Function;
import java.util.function.Supplier;

public abstract class Identification {

    // Each identification type is created by its own class, looked up by the type read from the record,
    // so adding a new type does not change this class more than this map.
    private static final Map<String, Function<String, Identification>> creatorPerType = Map.of(
        DNI.TYPE, DNI::new,
        CUIT.TYPE, CUIT::new);

    private final String number;

    // instance creation

    public static Identification ofType(String anIdentificationType, String anIdentificationNumber) {
        return Optional.ofNullable(creatorPerType.get(anIdentificationType))
            .orElseThrow(() -> new RuntimeException(invalidIdentificationTypeErrorDescription()))
            .apply(anIdentificationNumber);
    }

    // error messages

    public static String invalidIdentificationTypeErrorDescription() {
        return "Invalid identification type";
    }

    // initialization

    protected Identification(String anIdentificationNumber) {
        number = anIdentificationNumber;
    }

    // identification

    public boolean isOfTypeAndNumber(String anIdentificationType, String anIdentificationNumber) {
        return type().equals(anIdentificationType) && number.equals(anIdentificationNumber);
    }

    public String number() {
        return number;
    }

    public abstract String type();

    // testing

    public abstract boolean isCUIT();

    public abstract boolean isDNI();

    // numbers

    public abstract Object cuitNumberIfNone(Supplier<Object> aNoneBlock);

    public abstract Object dniNumberIfNone(Supplier<Object> aNoneBlock);
}
